File: DCM_Matlab/DCM.m
function[DC_sets, iterations, time, meanCor1, meanCor2, indices] = DCM(M1, M2, max_groups, echos, resid_full, max_iter, max_time, est_size, start, strict)

    % Options: max_groups, max_iter, max_time, est_size, start, resid_full, echos, strict
    % defaults = [5, 50, 1, 100, [], false, true, 'low']
    
    % optional args
    if ~exist('max_groups', 'var')
        max_groups = 5;
    end
    
    if ~exist('max_iter', 'var')
        max_iter = 100;
    end
    
    if ~exist('max_time', 'var')
        max_time = 160;
    end
    
    if ~exist('est_size', 'var')
        est_size = 100;
    end
    
    if ~exist('start', 'var')
        start = [];
    end
    
    if ~exist('resid_full', 'var')
       resid_full = false;
    end
    
    if ~exist('echos', 'var')
        echos = true;
    end
    
    if ~exist('strict', 'var')
        strict = 'low';
    end
    
    % In case we find nothing, initialize
    DC_sets = {};
    iterations = [];
    time = [];
    meanCor1 = [];
    meanCor2 = [];
    indices = [];

    alpha = 0.05;

    [M1, M2, idcs] = prepData_DCM(M1, M2, resid_full, echos, strict);
    
    tic;
    
    ans = 1;
    kval = est_size;
    startdels = [];
    p = size(M1, 1);
    
    while length(startdels) < p-kval & ans <= max_groups & toc < max_time
        
        % Initialize
        if length(start) < 5
            tmp = init_DCM(M1, M2, kval, startdels);
        else
            tmp = start;
            start = [];
        end
        
        if echos
            disp('Initialized');
            %disp(tmp);
        end
        
        diag = false;
        
        % Run search procedure
        [m_found, m_meanA1, m_meanA2, m_it, m_time, m_test, m_startdels] = run_DCM(M1, M2, tmp, startdels, echos, max_iter, alpha);

        startdels = [startdels, m_startdels];
       
		res = m_found;
		
		k = length(res);
		
		% Ignore groups that are too small to be significant
		if k > 10
			
			% Check for diag blocks
			if res(1) == 0
				diag = true;
				cut = res(2);
				res([1,2]) = [];
            end
			
			if diag
				res1 = res(1:cut);
				res2 = res((cut+1):end);
				
				% Find average correlations of result within each group
				crossCor1 = mean(mean(M1(res1, :) * M1(res2, :).'));
                crossCor2 = mean(mean(M2(res1, :) * M2(res2, :).'));
				
				if echos
					disp(sprintf('Found an off-diagonal block group of size %i', k));
					disp(sprintf('cross cor 1 = %f', crossCor1));
					disp(sprintf('cross cor 2 = %f', crossCor2));
                end

                 %startdels = [startdels, res1, res2];
				
				
            else
			
				% Announce result
				if echos
					disp(sprintf('Found a group of size %i', k));
					disp(sprintf('cor 1 = %f', m_meanA1));
					disp(sprintf('cor 2 = %f', m_meanA2));
                end

                 %startdels = [startdels, res];

            end
			
			% Residualize remaining data for continued search
			M1(res, :) = resid_DCM(M1(res, :));
			M2(res, :) = resid_DCM(M2(res, :));
			
			% Save results
			if diag
				DC_sets{ans} = {'Block Diag', idcs(res1), idcs(res2)};
				meanCor1{ans} = crossCor1;
				meanCor2{ans} = crossCor2;
            else
				DC_sets{ans} = idcs(res);
				meanCor1{ans} = m_meanA1;
				meanCor2{ans} = m_meanA2;
            end
			iterations{ans} = m_it;
			time{ans} = m_time;

			
			% Count number of sets found
			ans = ans+1;
			
        else
			
			% If group is degenerate, do not try it again

			startdels = [startdels, tmp.'];
			
        end % if(k > 10)
		
		% Check total runtime
		tottime = toc;
		
    end % while(length(dels) < p1-kval & ans < max_groups)
	
	% disp reason for algorithm halting
	if echos
		if toc >= max_time
			disp('Timed out.');
        elseif(length(startdels) > p-kval)
			disp('Exhausted all searches.');
        elseif(ans > max_groups)
			disp(sprintf('%i groups found.', max_groups));
        end
    end
	
	% Return results and properties each result
end
       

File: DCM/R/DCM.R
DCM <-
function(M1, M2, max.groups = 5, max.iter = 50, max.time = 1, est.size = nrow(M1)/10, alpha = 0.05, start = c(), QN = FALSE, resid.full = FALSE, echo = FALSE, strict = "low"){
	
	################### Prepare Data #####################
	
	prep = prepData_DCM(M1, M2, QN, resid.full, echo, strict)
	M1 = prep$M1
	M2 = prep$M2
	idcs = prep$idcs
	
	p = nrow(M1)
	
	################### Run Algorithm ####################

	# Prepare to store results
	DC_sets = list()
	iterations = list()
	time = list()
	meanCor1 = list()
	meanCor2 = list()


	# Initialize
	starttime = Sys.time()
	tottime = 0
	ans = 1
	kval = est.size
	startdels = c()
	
	# Keep searching until all rows or exhausted, or max groups or time is reached
	while(length(startdels) < p-kval & ans <= max.groups & tottime < max.time){
		
		# Find initial starting set, ignoring already tried starting points
		if(length(start) < 5){
			tmp = init_DCM(M1, M2, k=kval, del = startdels)$found
		}else{
			tmp = start
			start = c()
		}
	
		if(echo){
			print("Initialized")
			print(tmp)
		}
		
		diag = FALSE
		
		# Run search procedure with specified values
		DCM = run_DCM(M1, M2, seed = tmp, echo = echo, max.iter = max.iter)
		
		startdels = c(startdels, DCM$startdels)

		res = DCM$found
		
		k = length(res)
		
		# Ignore groups that are too small to be significant
		if(k > 10){
			
			# Check for diag blocks
			if(res[1] == 0){
				diag = TRUE
				cut = res[2]
				res = res[-c(1,2)]
			}
			
			if(diag){
				res1 = res[1:cut]
				res2 = res[-(1:cut)]
				
				# Find average correlations of result within each group
				### Fix this for length 1
				#crossCor1 = mean(cor(t(M1[res1,]), t(M1[res2,])))
				#crossCor2 = mean(cor(t(M2[res1,]), t(M2[res2,])))
				
				if(echo){
					print(sprintf("Found an off-diagonal block group of size %i", k))
					#print(sprintf("cross cor 1 = %f", crossCor1))
					#print(sprintf("cross cor 2 = %f", crossCor2))
				}
				
				
			}else{
			
				# Announce result
				if(echo){
					print(sprintf("Found a group of size %i", k))
					print(sprintf("cor 1 = %f", DCM$mc1))
					print(sprintf("cor 2 = %f", DCM$mc2))
				}
			
			}
			
			# Residualize remaining data for continued search
			M1[res,] = resid_DCM(M1[res,], QN = QN)
			M2[res,] = resid_DCM(M2[res,], QN = QN)
			
			# Save results
			if(diag){
				DC_sets[[ans]] = list("Block Diag", idcs[res1], idcs[res2])
				#meanCor1[[ans]] = crossCor1
				#meanCor2[[ans]] = crossCor2				
				meanCor1[[ans]] = 0
				meanCor2[[ans]] = 0
			}else{
				DC_sets[[ans]] = idcs[res]
				meanCor1[[ans]] = DCM$mc1
				meanCor2[[ans]] = DCM$mc2
			}
			iterations[[ans]] = DCM$its
			time[[ans]] = DCM$time

			
			# Count number of sets found
			ans = ans+1
			
		}else{
			
			# If group is degenerate, do not try it again
			startdels = c(startdels, tmp)
			
		}# if(k > 10)
		
		# Check total runtime
		tottime = difftime(Sys.time(), starttime, units = "hours")
		
	} # while(length(dels) < p-kval & ans < max.groups)
	
	# Print reason for algorithm halting
	if(echo){
		if(tottime >= max.time){
			print("Timed out.")
		}else if(length(startdels) > p-kval){
			print("Exhausted all searches.")
		}else if(ans > max.groups){
			print(sprintf("%i groups found.", max.groups))
		}
	}
	
	# Return results and properties each result
	return(list(DC_sets = DC_sets, iterations = iterations, time = time, meanCor1 = meanCor1, meanCor2 = meanCor2, indices = idcs))

}
